alerts/handler.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2017 alertmanager-github-receiver Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//////////////////////////////////////////////////////////////////////////////
+package alerts
+
+import (
+	"encoding/json"
+	"github.com/google/go-github/github"
+	//"github.com/kr/pretty"
+	"github.com/prometheus/alertmanager/notify"
+	"io/ioutil"
+	"log"
+	"net/http"
+)
+
+type ReceiverClient interface {
+	CloseIssue(issue *github.Issue) (*github.Issue, error)
+	CreateIssue(title, body string) (*github.Issue, error)
+	ListOpenIssues() ([]*github.Issue, error)
+}
+
+type ReceiverHandler struct {
+	Client ReceiverClient
+}
+
+// ServeHTTP receives and processes alertmanager notifications. If the alert
+// is firing and a github issue does not yet exist, one is created. If the
+// alert is resolved and a github issue exists, then it is closed.
+func (rh *ReceiverHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	// Verify that request is a POST.
+	if req.Method != http.MethodPost {
+		log.Printf("Client used unsupported method: %s: %s", req.Method, req.RemoteAddr)
+		rw.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Read request body.
+	alertBytes, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		log.Printf("Failed to read request body: %s", err)
+		rw.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// The WebhookMessage is dependent on alertmanager version. Parse it.
+	msg := &notify.WebhookMessage{}
+	if err := json.Unmarshal(alertBytes, msg); err != nil {
+		log.Printf("Failed to parse webhook message from %s: %s", req.RemoteAddr, err)
+		log.Printf("%s", string(alertBytes))
+		rw.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	// log.Print(pretty.Sprint(msg))
+
+	// Handle the webhook message.
+	log.Printf("Handling alert: %s", id(msg))
+	if err := rh.processAlert(msg); err != nil {
+		log.Printf("Failed to handle alert: %s: %s", id(msg), err)
+		rw.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	log.Printf("Completed alert: %s", id(msg))
+	rw.WriteHeader(http.StatusOK)
+	// Empty response.
+}
+
+// processAlert processes an alertmanager webhook message.
+func (rh *ReceiverHandler) processAlert(msg *notify.WebhookMessage) error {
```

Comment:
Good idea. I'd like to make this a TODO also right now.

Today our alertmanager configuration only fires on unique `alertname`s. This is to limit the set of issues possibly created at one time until we can safely create an issue for every "alertname, instance" pair. 

And, based on our discussion yesterday, I think our near term preference is to use labels more than "open" as the only list criteria. This may change the kinds of queries we want to cache and their cost.